---
title: Welcome to Jurassic Park - A tour of Ibex.
author: 
- name: Nick Borcherding
  email: ncborch@gmail.com
  affiliation: Washington University in St. Louis, School of Medicine, St. Louis, MO, USA
date: "July 29,2022"
output:
  BiocStyle::html_document:
    toc_float: true
package: Ibex
vignette: >
  %\VignetteEngine{knitr::knitr}
  %\VignetteIndexEntry{Charging through Ibex}
  %\usepackage[UTF-8]{inputenc}
---

```{r, echo=FALSE, results="hide", message=FALSE}
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
library(BiocStyle)
```

# Introduction

## Load Libraries

```{r}
suppressMessages(library(Seurat))
suppressMessages(library(ggplot2))
suppressMessages(library(Ibex))
suppressMessages(library(viridis))
suppressMessages(library(patchwork))
```


## The Data Set

To show the multiple options of Ibex, the example data is derived from [GSE167118](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE167118) - a cohort of CITE-seq data derived from severe COVID-19 patients. The [corresponding manuscript](https://pubmed.ncbi.nlm.nih.gov/33622974/) is excellent. The data example built into the package (ibex_example) is derived from randomly sampling the single-cells from patient 17. 

This is a standard workflow based on the [WNN](https://satijalab.org/seurat/articles/weighted_nearest_neighbor_analysis.html) Seurat process. However, Ibex will work for Bioconductor/Single-Cell Experiment workflows as well. The major exception is the addition of ```quietBCRgenes()```, which removed B cell receptor genes from the variable gene list used for ```runPCA()```. 


```{r, eval=FALSE}
suppressMessages(library(scRepertoire))
##################################
#scRNA/ADT loading and processing
#################################
tmp <-  Read10X("~/Patient17/filtered_feature_bc_matrix")

Pt17 <- CreateSeuratObject(counts = tmp$`Gene Expression`)
Pt17 <- subset(Pt17, subset = nFeature_RNA > 100) 
Pt17  <- RenameCells(object = Pt17 , new.names = paste0("Pt17_", rownames(Pt17[[]])))
Pt17[["mito.genes"]] <- PercentageFeatureSet(Pt17, pattern = "^MT-")
    
#Filtering step
standev <- sd(log(Pt17$nFeature_RNA))*2.5 #cutting off above standard deviation of 2.5
mean <- mean(log(Pt17$nFeature_RNA))
cut <- round(exp(standev+mean))
Pt17 <- subset(Pt17, subset = mito.genes < 10 & nFeature_RNA < cut)

#Processing RNA
DefaultAssay(Pt17) <- 'RNA'
Pt17 <- NormalizeData(Pt17) %>% FindVariableFeatures() %>% 
  quietBCRgenes() %>% ScaleData() %>% RunPCA(verbose = FALSE)

#Processing ADT
DefaultAssay(Pt17) <- 'ADT'
VariableFeatures(Pt17) <- rownames(Pt17[["ADT"]])
Pt17 <- NormalizeData(Pt17, normalization.method = 'CLR', margin = 2) %>% 
  ScaleData() %>% RunPCA(reduction.name = 'apca')


##################################
#Processing and Adding Contig Info
##################################

contigs <- read.csv("~/Patient17/filtered_contig_annotations.csv")
clones <- combineBCR(contigs, samples = "Pt17", cells = "T-AB", filterMulti = TRUE, removeNA = TRUE)
Pt17 <- combineExpression(clones, Pt17, cloneCall="aa")
saveRDS(Pt17, file = "Ibex_FullExample.rds")

###################################
#Making Example Data Set for Ibex
#################################
meta <- Pt17[[]]
meta <- meta[sample(nrow(meta), nrow(meta)*0.1),]
ibex_example <- subset(Pt17, cells = rownames(meta))
save(ibex_example, file = "ibex_example.rda", compress = "xz")
```

## Loading the Data

```{r}
data("ibex_example")
```

****************

# Running Ibex

The idea behind Ibex is to combine BCR cdr3 amino acid information with phenotypic RNA/protein data to direct the use of single-cell sequencing towards antigen-specific discoveries.  This is a growing field - specifically [TESSA](https://github.com/jcao89757/TESSA) uses amino acid characteristics and autoencoder as a means to get a dimensional reduction. Another option is [CoNGA](https://github.com/phbradley/conga), which produces an embedding using BCR and RNA. Ibex was designed to make a customizable approach to this combined approach using R.

## Ibex.matrix Function

Ibex has 2 major functions - the first being ```Ibex.matrix()```, which is the backbone of the algorithm and returns the encoded values based on the selection of variables. Unlike ```runIbex()``` below, ```Ibex.matrix()``` does not filter the input for only B cells with attached BCR data. 

**chains**  
* "Heavy" for Ig Heavy Chain 
* "Light" for Ig Light Chain 

**AA.properties**  
* "AF" to use Atchley factors  
* "KF" to use Kidera factors  
* "both" to use both  
* "OHE" for a One Hot Autoencoder

```{r}
Ibex_vectors <- Ibex.matrix(ibex_example, 
                       chains = "Heavy",
                      AA.properties = "OHE")

qplot(data = as.data.frame(Ibex_vectors), Ibex_2, Ibex_3) + theme_classic()
```


## runIbex

Additionally, ```runIbex()``` can be used to append the Seurat or Single-cell Experiment object with the Ibex vectors and allow for further analysis. Importantly, ```runIbex()``` will remove single cells that do not have recovered BCR data in the metadata of the object. 

```{r}
ibex_example <- runIbex(ibex_example, 
                    chains = "TRB",
                    AA.properties = "KF", 
                    reduction.name = "Ibex.KF")
```

## Using Ibex Vectors

After ```runIbex()``` we have the encoded values stored under **"Ibex..."**. Using the Ibex reduction stored in Seurat, we can calculate the nearest neighbor and shared nearest neighbor indexes and generate a UMAP.

```{r}
#Generating UMAP from Ibex Neighbors
ibex_example <- RunUMAP(ibex_example, 
                        reduction = "Ibex.KF",
                        dims = 1:30,
                        reduction.name = 'Ibex.umap', 
                        reduction.key = 'ibexUMAP_')
#Ibex UMAP
plot1 <- DimPlot(ibex_example, reduction = "Ibex.umap") + NoLegend()
plot2 <- DimPlot(ibex_example, group.by = "CTaa", reduction = "Ibex.umap") + 
  scale_color_viridis(discrete = TRUE, option = "B") + 
  NoLegend()

plot1 + plot2
```


We now can use this in a similar way as other single-cell modalities and calculate weighted nearest neighbor (WNN). To check out more on WNN, please read the Satija's group [paper](https://pubmed.ncbi.nlm.nih.gov/34062119/). We will use the RNA, ADT protein levels, and Ibex vectors for the WNN calculations. 

```{r}
ibex_example <- FindMultiModalNeighbors(
  ibex_example, reduction.list = list("pca", "apca", "Ibex.KF"), 
  dims.list = list(1:30, 1:20, 1:30), modality.weight.name = "RNA.weight"
)
ibex_example <- RunUMAP(ibex_example, 
                        nn.name = "weighted.nn", 
                        reduction.name = "wnn.umap", 
                        reduction.key = "wnnUMAP_")

ibex_example <- FindClusters(ibex_example, 
                             graph.name = "wsnn",
                             resolution = 0.6,
                             algorithm = 3, 
                             verbose = FALSE)

#WNN UMAP
plot3 <- DimPlot(ibex_example, reduction = "wnn.umap")
plot4 <- DimPlot(ibex_example, reduction = "wnn.umap", group.by = "CTaa") + 
  scale_color_manual(values = viridis_pal(option = "B")(length(unique(ibex_example$CTaa)))) + 
  NoLegend()

plot3 + plot4
```

## Comparing the outcome to just one modality

We can also look at the differences in the UMAP generated from RNA, ADT, or Ibex as individual components. Remember, the clusters that we are displaying in UMAP are based on clusters defined by the weighted nearest neighbors calculated above. 

```{r}
ibex_example <- RunUMAP(ibex_example, 
                        reduction = 'pca', 
                        dims = 1:30, 
                        assay = 'RNA', 
                        reduction.name = 'rna.umap', 
                        reduction.key = 'rnaUMAP_')

ibex_example <- RunUMAP(ibex_example, 
                        reduction = 'apca', 
                        dims = 1:20, 
                        assay = 'ADT', 
                        reduction.name = 'adt.umap', 
                        reduction.key = 'adtUMAP_')

plot5 <- DimPlot(ibex_example, reduction = "rna.umap") + NoLegend()
plot6 <- DimPlot(ibex_example, reduction = "adt.umap") + NoLegend()
plot7 <- DimPlot(ibex_example, reduction = "Ibex.umap") + NoLegend()

plot5 + plot6 + plot7
```

## Regressing shared feature space

There is a high degree of overlap in the feature space of expanded clonotypes, not only in the adaptive receptor genes that form the clonotype, but also in potentially functional genes. To identify divergent clusters in shared clonotypes using ```clonalCommunity()```, we can use the Ibex dimensions to identify clusters in order to then regress out closely associated clonotypes.

```{r}
library(harmony)
library(bluster)

#Cluster clones
ibex_example <- clonalCommunity(ibex_example, 
                            reduction.name = "Ibex.KF", 
                            cluster.parameter=NNGraphParam(k=20))

#Run Harmony and new UMAP for RNA
DefaultAssay(ibex_example) <- "RNA"
ibex_example<- RunHarmony(ibex_example, "ibex.clusters", verbose = FALSE)
ibex_example <- RunUMAP(ibex_example, 
                     reduction = "harmony", 
                     dims = 1:20, 
                     reduction.key = "RC_UMAP_", 
                     reduction.name = "regressClone_RNA")


plot8 <- DimPlot(ibex_example, 
                 reduction = "rna.umap", 
                 group.by = "CTaa") + 
                 NoLegend() + 
                 scale_color_viridis(option="B", discrete = TRUE)

plot9 <- DimPlot(ibex_example, 
                 reduction = "regressClone_RNA", 
                 group.by = "CTaa") + 
                 NoLegend() + 
                 scale_color_viridis(option="B", discrete = TRUE)

plot8 + plot9
```

# CoNGA Reduction

Recent [work](https://pubmed.ncbi.nlm.nih.gov/34426704/) has proposed using representative cells for the characterization of clonotype and gene expression relationships. In order to generate these representative cells, either a mean expression across a clone or using the PCA dimensional space to identify a single cell that has the minimum euclidean distance across a clone. 

In order to generate a single-cell object based on the CoNGA approach, Ibex offers the function ```CoNGAfy()```. For **method**, select either "mean" or "dist" as described above. After performing ```CoNGAfy()```, the user can use any of the above reduction strategies. 

```{r}
CoNGA.seurat <- CoNGAfy(ibex_example, 
                         method = "dist")

CoNGA.seurat <- runIbex(CoNGA.seurat, 
                    AA.properties = "KF", 
                    reduction.name = "Ibex.KF")

CoNGA.seurat <- CoNGA.seurat %>%
                  FindNeighbors(reduction = "Ibex.KF") %>%
                  FindClusters(algorithm = 3, resolution = 0.9)

CoNGA.seurat <- RunUMAP(CoNGA.seurat, 
                     reduction = "Ibex.KF", 
                     dims = 1:20, 
                     reduction.name = 'Ibex.umap', 
                     reduction.key = 'ibexUMAP_')

DimPlot(CoNGA.seurat, reduction = "Ibex.umap") + NoLegend()
```

# Conclusion

This has been a general overview of the capabilities of Ibex for incorporating BCR information into the embedding space of single-cell data. If you have any questions, comments, or suggestions, feel free to visit the github repository or [email me](mailto:ncborch@gmail.com).

## Session Info

```{r}
sessionInfo()
```
