#' Ibex Matrix Interface
#'
#' This function runs the Ibex algorithm to generate latent vectors from input data.
#' The output can be returned as a matrix, with options to choose between deep learning 
#' autoencoders or geometric transformations based on the BLOSUM62 matrix.
#'
#' @examples
#' # Using the encoder method with a variational autoencoder
#' ibex_values <- Ibex.matrix(ibex_example, 
#'                            chain = "Heavy",
#'                            method = "encoder",
#'                            encoder.model = "VAE",
#'                            encoder.input = "atchleyFactors")
#'
#' # Using the geometric method with a specified angle
#' ibex_values <- Ibex.matrix(ibex_example, 
#'                            chain = "Heavy",
#'                            method = "geometric",
#'                            geometric.theta = pi)
#'
#' @param input.data Input data, which can be:
#'   \itemize{
#'     \item A Single Cell Object in Seurat or SingleCellExperiment format
#'     \item The output of \code{combineBCR()} from the \code{scRepertoire} package
#'   }
#' @param chain Character. Specifies which chain to analyze:
#'   \itemize{
#'     \item "Heavy" for the heavy chain
#'     \item "Light" for the light chain
#'   }
#' @param method Character. The algorithm to use for generating latent vectors:
#'   \itemize{
#'     \item "encoder" - Uses deep learning autoencoders
#'     \item "geometric" - Uses geometric transformations based on the BLOSUM62 matrix
#'   }
#' @param encoder.model Character. The type of autoencoder model to use:
#'   \itemize{
#'     \item "CNN" - Convolutional Neural Network-based autoencoder
#'     \item "VAE" - Variational Autoencoder
#'   }
#' @param encoder.input Character. Specifies the input features for the encoder model. Options include:
#'   \itemize{
#'     \item Amino Acid Properties: "atchleyFactors", "crucianiProperties", "FASGAI", "kideraFactors", "MSWHIM", "ProtFP", "stScales", "tScales", "VHSE", "zScales"
#'     \item "OHE" for One Hot Encoding (mandatory for biLSTM models)
#'   }
#' @param geometric.theta Numeric. Angle (in radians) for the geometric transformation. Only used when \code{method = "geometric"}.
#'
#' @return A matrix of latent vectors generated by the specified method.
#'
#' @export
#' @importFrom SeuratObject CreateDimReducObject
#' @importFrom immApex propertyEncoder onehotEncoder geometricEncoder getIR
#'
#' @seealso 
#' \code{\link[immApex]{propertyEncoder}}, 
#' \code{\link[immApex]{geometricEncoder}}
Ibex.matrix <- function(input.data, 
                        chain = "Heavy", 
                        method = "encoder",
                        encoder.model = "VAE", 
                        encoder.input = "AF",
                        geometric.theta = pi/3) {
    #Chain Check
    if(chains %!in% c("Heavy", "Light")) {
      stop("Please select one of the following chains: 'Heavy', 'Light'")
    }
  
    #Model Check
    valid.encoder.inputs <- c("atchleyFactors", "crucianiProperties", "kideraFactors", "MSWHIM", "tScales", "OHE")
    if(encoder.input %!in% valid.encoder.inputs) {
      stop("Please select one of the valid encoder inputs.")
    }
  
    #Will be used to filter output of getIR()
    loci <- ifelse(chain == "Heavy", "IGH", c("IGK", "IGL"))
    expanded.sequences <- grepl(".Exp", encoder.model)
    
    #Set Dictionary for embedding
    if(expanded.sequences) {
      dictionary <- c(amino.acids, "_")
    } else {
      dictionary <- amino.acids
    }
    
    #Getting Sequences
    BCR <- getIR(input.data, chain, sequence.type = "aa")[[1]]
    
    #Filtering out NA values 
    if(any(is.na(BCR[,2]))) {
      BCR <- BCR[-which(is.na(BCR[,2])),]
    }
    #Filtering out sequences that do not match gene locid
    if(any(!grep(paste0(loci, collapse = "|"), BCR[,"v"]))) {
      BCR <- BCR[-!grep(paste0(loci, collapse = "|"), BCR[,"v"]),]
    }
    
    #Checking Sequences - needs to be < 45 or 90
    checkLength(x = BCR[,2], expanded = expanded.sequences)
    length.to.use <- ifelse(expanded.sequences, 90, 45)
    
    
    if (method == "encoder") {
      print("Encoding Sequences...")
      
      if(encoder.input == "OHE") {
          encoded.values <- suppressMessages(onehotEncoder(BCR[,2],
                                                           max.length = length.to.use,
                                                           convert.to.matrix = TRUE,
                                                           sequence.dictionary = dictionary,
                                                           padding.symbol = "."))
      } else {
          encoded.values <- suppressMessages(propertyEncoder(BCR[,2], 
                                                             max.length = length.to.use,
                                                             method.to.use = encoder.input,
                                                             convert.to.matrix = TRUE))
      }
      print("Calculating Latent Dimensions...")
      #Getting Model
      aa.model <- aa.model.loader(species = "Human", chain, encoder.input, encoder.model)
      reduction <- stats::predict(aa.model, encoded.values, verbose = 0)
    } else if (method == "geometric") {
        print("Performing geometric transformation...")
        reduction <- suppressMessages(geometricEncoder(BCR, theta = geometric.theta))
    }
    #TODO Check the barcode usage here
    reduction <- as.data.frame(reduction)
    barcodes <- reduction[,1]
    reduction <- reduction[,-1]
    rownames(reduction) <- barcodes
    colnames(reduction) <- paste0("Ibex_", seq_len(ncol(reduction)))
    return(reduction)
}

#' Ibex Single-Cell Calculation
#'
#' This function applies the Ibex algorithm to single-cell data, integrating seamlessly with 
#' Seurat or SingleCellExperiment pipelines. The algorithm generates latent dimensions using 
#' deep learning or geometric transformations, storing the results in the dimensional reduction slot.
#'
#' @examples
#' # Using the encoder method with a variational autoencoder
#' ibex_example <- runIbex(ibex_example, 
#'                         chain = "Heavy",
#'                         method = "encoder",
#'                         encoder.model = "VAE",
#'                         encoder.input = "atchleyFactors")
#'
#' # Using the geometric method with a specified angle
#' ibex_example <- runIbex(ibex_example, 
#'                         chain = "Heavy",
#'                         method = "geometric",
#'                         geometric.theta = pi)
#'
#' @param sc.data A single-cell dataset, which can be:
#'   \itemize{
#'     \item A Seurat object
#'     \item A SingleCellExperiment object
#'   }
#' @param chain Character. Specifies the chain to analyze:
#'   \itemize{
#'     \item "Heavy" for the heavy chain
#'     \item "Light" for the light chain
#'   }
#' @param method Character. Algorithm to use for generating latent dimensions:
#'   \itemize{
#'     \item "encoder" - Uses deep learning autoencoders
#'     \item "geometric" - Uses geometric transformations based on the BLOSUM62 matrix
#'   }
#' @param encoder.model Character. Type of autoencoder model to use:
#'   \itemize{
#'     \item "CNN" - Convolutional Neural Network-based autoencoder
#'     \item "VAE" - Variational Autoencoder
#'   }
#' @param encoder.input Character. Input features for the encoder model:
#'   \itemize{
#'     \item Amino Acid Properties: "atchleyFactors", "crucianiProperties", "kideraFactors", "MSWHIM", "tScales"
#'     \item "OHE" - One Hot Encoding 
#'   }
#' @param geometric.theta Numeric. Angle (in radians) for geometric transformation. 
#'   Used only when \code{method = "geometric"}.
#' @param reduction.name Character. The name to assign to the dimensional reduction. 
#'   This is useful for running Ibex with multiple parameter settings and saving results 
#'   under different names.
#'
#' @return An updated Seurat or SingleCellExperiment object with Ibex dimensions added 
#' to the dimensional reduction slot.
#' @export
runIbex <- function(sc.data, 
                    chains = "Heavy", 
                    method = "encoder",
                    encoder.model = "VAE", 
                    encoder.input = "AF",
                    geometric.theta = pi,
                    reduction.name = "Ibex") {
    checkSingleObject(sc.data)
    cells.chains <- rownames(sc.data[[]][!is.na(sc.data[["CTaa"]]),])
    sc.data <- subset(sc.data, cells = cells.chains)
    reduction <- Ibex.matrix(input.data = sc.data,
                             chains = chains, 
                             method = method,
                             encoder.model = encoder.model, 
                             encoder.input = encoder.input,
                             geometric.theta = geometric.theta)
    BCR <- getIR(input.data, chain, sequence.type = "aa")[[1]]
    sc.data <- adding.DR(sc.data, reduction, reduction.name)
    return(sc.data)
}
